src/org/sosy_lab/common/collect/OrderStatisticSet.java

```diff
@@ -0,0 +1,150 @@
+ *  limitations under the License.
+ */
+package org.sosy_lab.common.collect;
+
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.NavigableSet;
+
+/**
+ * A {@link NavigableSet} that allows two additional operations: receiving (and deleting) an element
+ * by its <i>rank</i>, and getting the rank of an element.
+ *
+ * <p>Implementations should adhere to all contracts of the <code>NavigableSet</code> interface.
+ *
+ * <p>Implementing classes should provide two means for comparing elements:
+ *
+ * <ol>
+ *   <li>Using the natural ordering of the elements. In this case, all elements of the set have to
+ *       implement the {@link Comparable} interface.
+ *   <li>Using a {@link java.util.Comparator Comparator} to create an order over the elements of the
+ *       set.
+ * </ol>
+ *
+ * <p>In both cases, the used compare-method should be consistent with <code>equals</code>, i.e.,
+ * <code>compare(a, b) == 0  =&gt;  a.equals(b)</code>, so that the contract provided by {@link
+ * java.util.Set Set} is fulfilled. If the used compare-method is not consistent with <code>equals
+ * </code>, the Set contract is not fulfilled.
+ *
+ * @param <E> the type of elements maintained by this set
+ */
+public interface OrderStatisticSet<E> extends NavigableSet<E> {
+
+  /**
+   * Returns the element of this set with the given rank. The lowest element in the set has rank ==
+   * 0, the largest element in the set has rank == size - 1.
+   *
+   * @param pIndex the rank of the element to return
+   * @return the element of this set with the given rank
+   * @throws IndexOutOfBoundsException if the given rank is out of the range of this set (i.e.,
+   *     pRank &lt; 0 || pRank &gt;= size)
+   */
+  E getByRank(int pIndex);
+
+  /**
+   * Remove the element of this set with the given rank and return it.
+   *
+   * <p>The lowest element in the set has rank == 0, the largest element in the set has rank == size
+   * - 1.
+   *
+   * @param pIndex the rank of the element to remove
+   * @return the removed element
+   * @throws IndexOutOfBoundsException if the given rank is out of the range of this set (i.e.,
+   *     pRank &lt; 0 || pRank &gt;= size)
+   * @see #getByRank(int)
+   */
+  @CanIgnoreReturnValue
+  E removeByRank(int pIndex);
+
+  /**
+   * Return the rank of the given element in this set. Returns -1 if the element does not exist in
+   * the set.
+   *
+   * <p>The lowest element in the set has rank == 0, the largest element in the set has rank == size
+   * - 1.
+   *
+   * @param pObj the element to return the rank for
+   * @return the rank of the given element in the set, or -1 if the element is not in the set
+   * @throws NullPointerException if the given element is <code>null</code>
+   */
+  int rankOf(E pObj);
+
+  @Override
+  OrderStatisticSet<E> descendingSet();
+
+  @Override
+  OrderStatisticSet<E> subSet(
```

Comment:
Btw., what is the rank of elements in a subset? Do the elements keep the same rank as in the original set, or do they get a new rank according to the contents of the subset? I would strongly prefer the latter. Are there tests for this?

The same question of course exists for all other subset views returned by some methods, and for the map methods. Furthermore, how does it work on the `descendingSet`?

The answers to these questions should be documented in the JavaDoc, either globally for all view-returning methods, or per method.

Reply:
I wouldn't even have thought of the first possibility.

`rankOf` and the other rank-methods work on the rank of the element in the `OrderStatistic*` collection the method is called on.
So if i call `subSet.rankOf(obj)`, it was obvious to me that, as the JavaDoc states, it returns the rank of `obj` in that specific set, i.e., `subSet`, not some backing set that the user may not even know of.

I have added clarifying Javadocs to the interface methods.